relayer/chain/ethereum/trie.go

```diff
@@ -2,26 +2,65 @@ package ethereum
 
 import (
 	"bytes"
+	"sync"
 
+	"github.com/ethereum/go-ethereum/common"
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/rlp"
 	"github.com/ethereum/go-ethereum/trie"
 )
 
 func MakeTrie(items types.Receipts) (*trie.Trie, error) {
-	keyBuf := new(bytes.Buffer)
+
 	trie := new(trie.Trie)
 
-	for i := 0; i < items.Len(); i++ {
-		keyBuf.Reset()
-		rlp.Encode(keyBuf, uint(i))
+	receiptTrie := CreateTrie(types.Receipts(items), trie)
+
+	return receiptTrie, nil
+}
+
+// Note - all code below is mostly copied from:
+// https://github.com/ethereum/go-ethereum/blob/85afdeef37e17bffd9d098e86312c569995705e0/core/types/hashing.go#L87
+
+// CreateTrie creates the tree hashes of transactions and receipts in a block header.
+func CreateTrie(list types.DerivableList, trie *trie.Trie) *trie.Trie {
+	trie.Reset()
 
-		value, err := rlp.EncodeToBytes(items[i])
-		if err != nil {
-			return nil, err
-		}
+	valueBuf := encodeBufferPool.Get().(*bytes.Buffer)
+	defer encodeBufferPool.Put(valueBuf)
 
-		trie.Update(keyBuf.Bytes(), value)
+	// StackTrie requires values to be inserted in increasing hash order, which is not the
```

Comment:
I'm using a normal Trie instead of a StackTrie because we need Trie.Prove() later on, which StackTrie does not have. I think StackTrie is just an optimization that prunes, and is not actually structured any differently, but I'm not 100% sure - https://github.com/ethereum/go-ethereum/pull/21407/files#diff-450c04b40c4e2f8c12ed10930cd0a8b8a6c25b77b1f3469e7205b43b5fad058aR46 has the implementation/further details to confirm

Reply:
Quickly glanced through the StackTrie code, looks indeed like it's only used as an optimization.